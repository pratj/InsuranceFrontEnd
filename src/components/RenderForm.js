import React, { useState } from "react";
import { useForm } from "react-hook-form";
import { makeStyles } from "@material-ui/core/styles";
import { Button, Container, TextField, Typography } from "@material-ui/core";
import "../RenderForm.css";
import {
  FormControl,
  FormLabel,
  RadioGroup,
  FormControlLabel,
  Radio,
  Switch,
} from "@material-ui/core";
import "date-fns";
import DateFnsUtils from "@date-io/date-fns";
import {
  MuiPickersUtilsProvider,
  KeyboardDatePicker,
} from "@material-ui/pickers";
import Controls from "../components/formControls/Controls/Controls";
const useStyles = makeStyles((theme) => ({
  paper: {
    marginTop: theme.spacing(2),
    display: "flex",
    flexDirection: "column",
    color: "#b58500",
  },
  form: {
    width: "100%",
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
    backgroundColor: "gold",
  },
}));

function RenderForm({ formFields, onSubmit }) {
  var validations;
  const classes = useStyles();
  const { register, handleSubmit, errors } = useForm();
  const [selectedStartDate, handleStartDateChange] = useState(new Date());
  const [selectedEndDate, handleEndDateChange] = useState(new Date());
  const handlePattern = (validation) => {
    let tempValidation = validation;
    tempValidation["pattern"]["value"] = new RegExp(validation.pattern.value);
    validations = tempValidation;
  };

  const handleValidations = (validation) => {
    validations = validation;
  };

  const renderFields = (fields) => {
    return fields.map((field) => {
      const { name, type, label, placeholder, validation } = field;
      {
        validation.hasOwnProperty("pattern")
          ? handlePattern(validation)
          : handleValidations(validation);
      }

      switch (type) {
        case "text":
        case "email":
        case "tel":
        case "number":
          return (
            <div key={name}>
              <TextField
                id={name}
                name={name}
                type={type}
                label={label}
                placeholder={placeholder}
                inputRef={register(validations)}
                variant="outlined"
                margin="normal"
                fullWidth
              />
              {errors[name] && (
                <span style={{ color: "red" }}>* {errors[name].message}</span>
              )}
            </div>
          );

        case "startdate":
          return (
            <div key={name}>
              <MuiPickersUtilsProvider utils={DateFnsUtils}>
                <KeyboardDatePicker
                  id={name}
                  name={name}
                  label={label}
                  value={selectedStartDate}
                  placeholder={placeholder}
                  variant="outlined"
                  format="MM/dd/yyyy"
                  margin="normal"
                  onChange={(date) => handleStartDateChange(date)}
                  minDate={new Date()}
                  InputLabelProps={{ shrink: true }}
                  inputRef={register(validations)}
                  fullWidth
                />
              </MuiPickersUtilsProvider>
              {errors[name] && (
                <span style={{ color: "red" }}>* {errors[name].message}</span>
              )}
            </div>
          );

        case "enddate":
          return (
            <div key={name}>
              <MuiPickersUtilsProvider utils={DateFnsUtils}>
                <KeyboardDatePicker
                  id={name}
                  name={name}
                  label={label}
                  value={selectedEndDate}
                  placeholder={placeholder}
                  variant="outlined"
                  format="MM/dd/yyyy"
                  margin="normal"
                  minDate={selectedStartDate}
                  onChange={(date) => handleEndDateChange(date)}
                  InputLabelProps={{ shrink: true }}
                  inputRef={register(validations)}
                  fullWidth
                />
              </MuiPickersUtilsProvider>

              {errors[name] && (
                <span style={{ color: "red" }}>* {errors[name].message}</span>
              )}
            </div>
          );

        case "select":
          return (
            <div key={name}>
              <select id={name} name={name} inputRef={register(validations)}>
                <option value="">Select Destination..</option>
                <option value="India">India</option>
                <option value="England">England</option>
                <option value="Australia">Australia</option>
                <option value="USA">USA</option>
              </select>
              {errors[name]  && (
                <span style={{ color: "red" }}>* {errors[name].message}</span>
              )}
              {/* <Controls.CountrySelect /> */}
              {/* <Controls.Asynchronous name={name} ref={register}/> */}
            </div>
          );

        case "radio":
          return (
            <div key={name}>
              <FormControl component="fieldset" required>
                <FormLabel component="legend">{label}</FormLabel>
                <RadioGroup row name={name}>
                  {field.items.map((item) => {
                    return (
                      <FormControlLabel
                        key={item.value}
                        value={item.value}
                        control={<Radio inputRef={register(validations)} />}
                        label={item.label}
                        labelPlacement="end"
                      />
                    );
                  })}
                </RadioGroup>
              </FormControl>
            </div>
          );

        default:
          return (
            <div key={name}>
              <span style={{ color: "red" }}>Invalid Field</span>
            </div>
          );
      }
    });
  };

  return (
    <Container component="main" maxWidth="xs" className="renderForm">
      <div className={classes.paper}>
        <Typography variant="subtitle1">
          Provide the following details
        </Typography>
        <form className={classes.form} onSubmit={handleSubmit(onSubmit)}>
          {renderFields(formFields)}
          <FormControlLabel
            control={<Switch name="sendUpdates" inputRef={register} />}
            label="Get updates on Email"
          />
          <Button
            className="button"
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
            style={{ borderRadius: "15px" }}
          >
            Submit
          </Button>
        </form>
      </div>
    </Container>
  );
}

export default RenderForm;
