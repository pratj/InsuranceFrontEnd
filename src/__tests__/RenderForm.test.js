import { shallow, mount } from "enzyme";
import RenderForm from "../components/RenderForm";
import { render, screen } from "@testing-library/react";
import toJson from "enzyme-to-json";

const onSubmit = (data) => {
  console.log(data);
};

const formFields = [
  {
    name: "fullName",
    type: "text",
    label: "Full Name",
    placeholder: "Enter Full Name",
    validation: {
      required: {
        value: true,
        message: "Full Name is required",
      },
    },
  },
];

test("renders renderForm component", () => {
  shallow(<RenderForm formFields={formFields} onSubmit={onSubmit} />);
});

test("renders renderForm component with parent div", () => {
  const wrapper = shallow(
    <RenderForm formFields={formFields} onSubmit={onSubmit} />
  );
  expect(wrapper.find('[data-test="renderForm"]').length).toBe(1);
});

test("renders `Provide the following details` heading", () => {
  render(<RenderForm formFields={formFields} onSubmit={onSubmit} />);
  expect(
    screen.getByText(/provide the following Details/i)
  ).toBeInTheDocument();
});

test("render a button with text of `Submit`", () => {
  const wrapper = shallow(
    <RenderForm formFields={formFields} onSubmit={onSubmit} />
  );
  expect(wrapper.find("form").text()).toBe("Submit");
});

test("accepts renderForm props", () => {
  const wrapper = mount(
    <RenderForm formFields={formFields} onSubmit={onSubmit} />
  );
  expect(wrapper.props().formFields).toEqual(formFields);
  expect(wrapper.props().onSubmit).toEqual(onSubmit);
});

test("RenderForm snapshot", () => {
  const renderFormTree = shallow(
    <RenderForm formFields={formFields} onSubmit={onSubmit} />
  );
  expect(toJson(renderFormTree)).toMatchSnapshot();
});