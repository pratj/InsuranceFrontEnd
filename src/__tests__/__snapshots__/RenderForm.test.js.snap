// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderForm snapshot 1`] = `
<WithStyles(ForwardRef(Container))
  className="renderForm"
  component="main"
  data-test="renderForm"
  maxWidth="xs"
>
  <div
    className="makeStyles-paper-1"
  >
    <WithStyles(ForwardRef(Typography))
      variant="subtitle1"
    >
      Provide the following details
    </WithStyles(ForwardRef(Typography))>
    <form
      className="makeStyles-form-2"
      onSubmit={[Function]}
    >
      <div
        key="fullName"
      >
        <WithStyles(ForwardRef(TextField))
          fullWidth={true}
          id="fullName"
          inputRef={[Function]}
          label="Full Name"
          margin="normal"
          name="fullName"
          placeholder="Enter Full Name"
          type="text"
          variant="outlined"
        />
      </div>
      <WithStyles(ForwardRef(FormControlLabel))
        control={
          <WithStyles(ForwardRef(Switch))
            inputRef={[Function]}
            name="sendUpdates"
          />
        }
        label="Get updates on Email"
      />
      <WithStyles(ForwardRef(Button))
        className="makeStyles-submit-3"
        color="primary"
        fullWidth={true}
        style={
          Object {
            "borderRadius": "15px",
          }
        }
        type="submit"
        variant="contained"
      >
        Submit
      </WithStyles(ForwardRef(Button))>
    </form>
  </div>
</WithStyles(ForwardRef(Container))>
`;
